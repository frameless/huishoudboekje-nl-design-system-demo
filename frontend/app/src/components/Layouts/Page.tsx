import {Box, Heading, Stack, StackProps} from "@chakra-ui/react";
import React from "react";

type PageProps = Omit<StackProps, "right"> & { title: string, backButton?: JSX.Element, menu?: JSX.Element, right?: JSX.Element };
const Page: React.FC<PageProps> = ({title, backButton, menu, right, children, ...props}) => {
	return (
		<Stack spacing={5} {...props}>
			{backButton && <Box>{backButton}</Box>}

			<Stack direction={"row"} justifyContent={"space-between"} alignItems={"center"}>
				<Stack direction={"row"} justifyContent={"flex-start"} alignItems={"center"} spacing={3}>
					<Heading size={"lg"}>{title}</Heading>
					{menu}
				</Stack>
				{right && <Box>{right}</Box>}
			</Stack>

			{children}
		</Stack>
	)
};

export default Page;