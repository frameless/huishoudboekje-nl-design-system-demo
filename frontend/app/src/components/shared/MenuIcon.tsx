import {createIcon} from "@chakra-ui/icon";

const MenuIcon = createIcon({
	// Taken from HiDotsVertical from react-icons/hi
	displayName: "MenuIcon",
	d: "M10 6a2 2 0 110-4 2 2 0 010 4z M10 12a2 2 0 110-4 2 2 0 010 4z M10 18a2 2 0 110-4 2 2 0 010 4z",
});

export default MenuIcon;