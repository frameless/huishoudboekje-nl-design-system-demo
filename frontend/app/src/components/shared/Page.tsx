import {Box, Heading, HStack, Stack} from "@chakra-ui/react";
import React, {ReactNode} from "react";

type PageProps = {
	title: string,
	backButton?: JSX.Element,
	menu?: JSX.Element | null | false,
	right?: JSX.Element | null | false,
	children?: ReactNode
};

const Page: React.FC<PageProps> = ({title, backButton, menu, right, children}) => {
	return (
		<Stack spacing={5}>
			{backButton && <Box>{backButton}</Box>}

			<Stack direction={"row"} justifyContent={"space-between"} alignItems={"start"}>
				<Heading className={"print"} size={"lg"}>{title}</Heading>
				<HStack>
					{right && <Box>{right}</Box>}
					{menu}
				</HStack>
			</Stack>

			{children}
		</Stack>
	);
};

export default Page;