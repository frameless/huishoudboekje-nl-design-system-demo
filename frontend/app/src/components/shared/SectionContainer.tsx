import {chakra, Divider, Stack, StackProps} from "@chakra-ui/react";
import React from "react";

const BaseSectionContainer = chakra(Stack, {
	baseStyle: {
		maxWidth: "100%",
		bg: "white",
		p: 5,
		borderRadius: 10,
	},
});

const SectionContainer: React.FC<StackProps> = (props) => (
	<BaseSectionContainer className="do-not-print" spacing={5} divider={<Divider />} {...props} />
)

export default SectionContainer;