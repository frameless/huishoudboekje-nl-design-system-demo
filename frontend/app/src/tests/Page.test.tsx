import {act} from "@testing-library/react";
import React from "react";
import {render, unmountComponentAtNode} from "react-dom";
import Page from "../components/shared/Page";
import {Button, ButtonGroup, IconButton, Menu, MenuButton, MenuItem, MenuList} from "@chakra-ui/react";
import MenuIcon from "../components/shared/MenuIcon";
import SectionContainer from "../components/shared/SectionContainer";

jest.mock("react-i18next", () => require("./utils/mock-hooks").reactI18NextMock());

let container: HTMLDivElement | null = null;

beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    unmountComponentAtNode(container!);
    container!.remove();
    container = null;
});

it("Page with only title", () => {
    act(() => {
        render(
            <Page
                title={"Bankafschriften"} />
            , container);
    });

    const html = container!.innerHTML;
    expect(html).not.toBeNull();
});

it("Page with title & backButton", () => {
    act(() => {
        render(
            <Page
                title={"Bankafschriften"}
                backButton={< Button />}
            />
            , container);
    });

    const html = container!.innerHTML;
    expect(html).not.toBeNull();
});

it("Page with title & menu", () => {
    act(() => {
        render(
            <Page
                title={"Bankafschriften"}
                menu={(
                    <Menu>
                        <IconButton as={MenuButton} icon={<MenuIcon />} variant={"solid"} aria-label={"Open menu"} />
                        <MenuList>
                            <MenuItem>Alle transacties afletteren</MenuItem>
                        </MenuList>
                    </Menu>
                )}
            />
            , container);
    });

    const html = container!.innerHTML;
    expect(html).not.toBeNull();
});

it("Page with title & right", () => {
    act(() => {
        render(
            <Page
                title={"Bankafschriften"}
                right={(
                    <ButtonGroup size={"sm"} isAttached variant={"outline"}>
                        <Button mr={"-px"} colorScheme={"primary"} variant={"outline"} size={"sm"}>Actie</Button>
                        <Button colorScheme={"primary"} variant={"outline"} size={"sm"}>Actie</Button>
                    </ButtonGroup>
                )}
            />
            , container);
    });

    const html = container!.innerHTML;
    expect(html).not.toBeNull();
});

it("Page with title & children", () => {
    act(() => {
        render(
            <Page title={"Bankafschriften"}>
                <SectionContainer />
            </Page>
            , container);
    });

    const html = container!.innerHTML;
    expect(html).not.toBeNull();
});

it("Page with all options", () => {
    act(() => {
        render(
            <Page
                title={"Bankafschriften"}
                backButton={<Button />}
                menu={(
                    <Menu>
                        <IconButton as={MenuButton} icon={<MenuIcon />} variant={"solid"} aria-label={"Open menu"} />
                        <MenuList>
                            <MenuItem>Alle transacties afletteren</MenuItem>
                        </MenuList>
                    </Menu>
                )}
                right={(
                    <ButtonGroup size={"sm"} isAttached variant={"outline"}>
                        <Button mr={"-px"} colorScheme={"primary"} variant={"outline"} size={"sm"}>Actie</Button>
                        <Button colorScheme={"primary"} variant={"outline"} size={"sm"}>Actie</Button>
                    </ButtonGroup>
                )}
            >
                <SectionContainer />
            </Page>
            , container);
    });

    const html = container!.innerHTML;
    expect(html).not.toBeNull();
});