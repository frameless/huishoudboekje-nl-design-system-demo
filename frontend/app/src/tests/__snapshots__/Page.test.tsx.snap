// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Renders a menu 1`] = `"<div class=\\"chakra-stack css-137fd6w\\"><div class=\\"chakra-stack css-1oeb4ru\\"><h2 class=\\"chakra-heading print css-0\\">Bankafschriften</h2><div class=\\"chakra-stack css-84zodg\\"><button class=\\"chakra-menu__menu-button chakra-button css-1bpn5qf\\" aria-label=\\"Open menu\\" id=\\"menu-button-:r4:\\" aria-expanded=\\"false\\" aria-haspopup=\\"menu\\" aria-controls=\\"menu-list-:r4:\\" type=\\"button\\"><span class=\\"css-xl71ch\\"><svg viewBox=\\"0 0 24 24\\" focusable=\\"false\\" class=\\"chakra-icon css-onkibi\\" aria-hidden=\\"true\\"><path fill=\\"currentColor\\" d=\\"M10 6a2 2 0 110-4 2 2 0 010 4z M10 12a2 2 0 110-4 2 2 0 010 4z M10 18a2 2 0 110-4 2 2 0 010 4z\\"></path></svg></span></button><div style=\\"visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;\\" class=\\"css-0\\"><div class=\\"chakra-menu__menu-list css-sjt5nk\\" tabindex=\\"-1\\" role=\\"menu\\" id=\\"menu-list-:r4:\\" style=\\"transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);\\" aria-orientation=\\"vertical\\"><button type=\\"button\\" id=\\"menu-list-:r4:-menuitem-:r5:\\" role=\\"menuitem\\" tabindex=\\"-1\\" class=\\"chakra-menu__menuitem css-1k756t9\\" data-index=\\"0\\">Alle transacties afletteren</button></div></div></div></div></div>"`;

exports[`Page Renders a menu 2`] = `"<div class=\\"chakra-stack css-137fd6w\\"><div class=\\"chakra-stack css-1oeb4ru\\"><h2 class=\\"chakra-heading print css-0\\">Bankafschriften</h2><div class=\\"chakra-stack css-84zodg\\"><button class=\\"chakra-menu__menu-button chakra-button css-1bpn5qf\\" aria-label=\\"Open menu\\" id=\\"menu-button-:r4:\\" aria-expanded=\\"true\\" aria-haspopup=\\"menu\\" aria-controls=\\"menu-list-:r4:\\" type=\\"button\\" data-active=\\"\\"><span class=\\"css-xl71ch\\"><svg viewBox=\\"0 0 24 24\\" focusable=\\"false\\" class=\\"chakra-icon css-onkibi\\" aria-hidden=\\"true\\"><path fill=\\"currentColor\\" d=\\"M10 6a2 2 0 110-4 2 2 0 010 4z M10 12a2 2 0 110-4 2 2 0 010 4z M10 18a2 2 0 110-4 2 2 0 010 4z\\"></path></svg></span></button><div style=\\"visibility: visible; position: absolute; min-width: max-content; inset: 0 auto auto 0; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 8px); --popper-transform-origin: top left;\\" class=\\"css-0\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"bottom-start\\"><div class=\\"chakra-menu__menu-list css-sjt5nk\\" tabindex=\\"-1\\" role=\\"menu\\" id=\\"menu-list-:r4:\\" style=\\"transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);\\" aria-orientation=\\"vertical\\"><button type=\\"button\\" id=\\"menu-list-:r4:-menuitem-:r5:\\" role=\\"menuitem\\" tabindex=\\"-1\\" class=\\"chakra-menu__menuitem css-1k756t9\\" data-index=\\"0\\">Alle transacties afletteren</button></div></div></div></div></div>"`;

exports[`Page Renders a title 1`] = `"<div class=\\"chakra-stack css-137fd6w\\"><div class=\\"chakra-stack css-1oeb4ru\\"><h2 class=\\"chakra-heading print css-0\\">Bankafschriften</h2><div class=\\"chakra-stack css-84zodg\\"></div></div></div>"`;

exports[`Page Renders component with children 1`] = `"<div class=\\"chakra-stack css-137fd6w\\"><div class=\\"chakra-stack css-1oeb4ru\\"><h2 class=\\"chakra-heading print css-0\\">Bankafschriften</h2><div class=\\"chakra-stack css-84zodg\\"></div></div><p class=\\"chakra-text css-0\\">Hier kunnen verschillende componenten komen te staan.</p></div>"`;

exports[`Page Renders components on the right 1`] = `"<div class=\\"chakra-stack css-137fd6w\\"><div class=\\"chakra-stack css-1oeb4ru\\"><h2 class=\\"chakra-heading print css-0\\">Bankafschriften</h2><div class=\\"chakra-stack css-84zodg\\"><div class=\\"css-0\\"><div class=\\"chakra-stack css-84zodg\\"><p class=\\"chakra-text css-0\\">Tekst zichtbaar</p><button type=\\"button\\" class=\\"chakra-button css-4xx2wk\\">Actie</button></div></div></div></div></div>"`;

exports[`Page Renders components on the right 2`] = `"<div class=\\"chakra-stack css-137fd6w\\"><div class=\\"chakra-stack css-1oeb4ru\\"><h2 class=\\"chakra-heading print css-0\\">Bankafschriften</h2><div class=\\"chakra-stack css-84zodg\\"><div class=\\"css-0\\"><div class=\\"chakra-stack css-84zodg\\"><p class=\\"chakra-text css-0\\">Tekst zichtbaar</p><button type=\\"button\\" class=\\"chakra-button css-4xx2wk\\">Actie</button></div></div></div></div></div>"`;

exports[`Page Renders the page component with all options 1`] = `"<div class=\\"chakra-stack css-137fd6w\\"><div class=\\"css-0\\"><button type=\\"button\\" class=\\"chakra-button css-4xx2wk\\">Click</button></div><div class=\\"chakra-stack css-1oeb4ru\\"><h2 class=\\"chakra-heading print css-0\\">Bankafschriften</h2><div class=\\"chakra-stack css-84zodg\\"><div class=\\"css-0\\"><div role=\\"group\\" class=\\"chakra-button__group css-1ayoy1w\\" data-attached=\\"\\" data-orientation=\\"horizontal\\"><button type=\\"button\\" class=\\"chakra-button css-swbvt8\\">Actie</button><button type=\\"button\\" class=\\"chakra-button css-1fysbuj\\">Actie</button></div></div><button class=\\"chakra-menu__menu-button chakra-button css-1bpn5qf\\" aria-label=\\"Open menu\\" id=\\"menu-button-:r1:\\" aria-expanded=\\"false\\" aria-haspopup=\\"menu\\" aria-controls=\\"menu-list-:r1:\\" type=\\"button\\"><span class=\\"css-xl71ch\\"><svg viewBox=\\"0 0 24 24\\" focusable=\\"false\\" class=\\"chakra-icon css-onkibi\\" aria-hidden=\\"true\\"><path fill=\\"currentColor\\" d=\\"M10 6a2 2 0 110-4 2 2 0 010 4z M10 12a2 2 0 110-4 2 2 0 010 4z M10 18a2 2 0 110-4 2 2 0 010 4z\\"></path></svg></span></button><div style=\\"visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;\\" class=\\"css-0\\"><div class=\\"chakra-menu__menu-list css-sjt5nk\\" tabindex=\\"-1\\" role=\\"menu\\" id=\\"menu-list-:r1:\\" style=\\"transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);\\" aria-orientation=\\"vertical\\"><button type=\\"button\\" id=\\"menu-list-:r1:-menuitem-:r2:\\" role=\\"menuitem\\" tabindex=\\"-1\\" class=\\"chakra-menu__menuitem css-1k756t9\\" data-index=\\"0\\">Alle transacties afletteren</button></div></div></div></div><div class=\\"chakra-stack css-r7hjxq\\"></div></div>"`;

exports[`Page Renders with a backbutton 1`] = `"<div class=\\"chakra-stack css-137fd6w\\"><div class=\\"css-0\\"><button type=\\"button\\" class=\\"chakra-button css-4xx2wk\\">Click</button></div><div class=\\"chakra-stack css-1oeb4ru\\"><h2 class=\\"chakra-heading print css-0\\">Bankafschriften</h2><div class=\\"chakra-stack css-84zodg\\"></div></div></div>"`;
